drivers/led/ht16k33.c

```diff
@@ -0,0 +1,282 @@
+/*
+ * Copyright (c) 2019 Henrik Brix Andersen <henrik@brixandersen.dk>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file
+ * @brief LED driver for the HT16K33 I2C LED driver with keyscan
+ */
+
+#include <i2c.h>
+#include <led.h>
+#include <zephyr.h>
+
+#define LOG_LEVEL CONFIG_LED_LOG_LEVEL
+#include <logging/log.h>
+LOG_MODULE_REGISTER(ht16k33);
+
+#include "led_context.h"
+
+/* HT16K33 commands and options */
+#define HT16K33_CMD_DISP_DATA_ADDR 0x00
+#define HT16K33_CMD_SYSTEM_SETUP   0x20
+#define   HT16K33_OPT_S            BIT(0)
+#define HT16K33_CMD_KEY_DATA_ADDR  0x40
+#define HT16K33_CMD_INT_FLAG_ADDR  0x60
+#define HT16K33_CMD_DISP_SETUP     0x80
+#define   HT16K33_OPT_D            BIT(0)
+#define   HT16K33_OPT_B0           BIT(1)
+#define   HT16K33_OPT_B1           BIT(2)
+#define     HT16K33_OPT_BLINK_OFF  0
+#define     HT16K33_OPT_BLINK_2HZ  HT16K33_OPT_B0
+#define     HT16K33_OPT_BLINK_1HZ  HT16K33_OPT_B1
+#define     HT16K33_OPT_BLINK_05HZ (HT16K33_OPT_B1 | HT16K33_OPT_B0)
+#define HT16K33_CMD_ROW_INT_SET    0xa0
+#define   HT16K33_OPT_ROW_INT      BIT(0)
+#define   HT16K33_OPT_ACT          BIT(1)
+#define     HT16K33_OPT_ROW        0
+#define     HT16K33_OPT_INT_LOW    HT16K33_OPT_ROW_INT
+#define     HT16K33_OPT_INT_HIGH   (HT16K33_OPT_ACT | HT16K33_OPT_ROW_INT)
+#define HT16K33_CMD_DIMMING_SET    0xe0
+
+/* HT16K33 size definitions */
+#define HT16K33_DISP_ROWS          16
+#define HT16K33_DISP_COLS          8
+#define HT16K33_DISP_DATA_SIZE     HT16K33_DISP_ROWS
+#define HT16K33_DISP_SEGMENTS      (HT16K33_DISP_ROWS * HT16K33_DISP_COLS)
+#define HT16K33_DIMMING_LEVELS     16
+
+struct ht16k33_cfg {
+	char *i2c_dev_name;
+	u16_t i2c_addr;
+};
+
+struct ht16k33_data {
+	struct device *i2c;
+	struct led_data dev_data;
+	u8_t buffer[HT16K33_DISP_DATA_SIZE];
```

Comment:
Can you clarify the use of this buffer?

Reply:
Comment added.

Reply:
I still don't understand the use of it to turn on/off the LEDs.

Reply:
@Mani-Sadhasivam It is a shadow buffer of the display RAM in the HT16K33. It's there to avoid an expensive I2C read-modify-write cycle when needing to turn on a single LED (and preserving the state of the remaining 7 LEDs, whose state is stored in the same byte of display RAM).

It will also be needed once I add Power Management support to the driver (pending). Then the LED states will be reinitialised with the date stored in that buffer upon resume from power-down.